pkg/cli/admin/release/new.go

```diff
@@ -810,6 +810,24 @@ func resolveImageStreamTagsToReferenceMode(inputIS, is *imageapi.ImageStream, re
 				continue
 			}
 
+			statusRef := findStatusTagEvents(inputIS.Status.Tags, ref.Name)
+
+			generation := int64(0)
+			if ref.Generation != nil {
+				generation = *ref.Generation
+				if statusRef != nil && len(statusRef.Items) > 0 {
+					newest := statusRef.Items[0]
+					// spec tags that are waiting for import should prevent release building,
+					if generation > newest.Generation {
+						return fmt.Errorf("the tag %q in the source input stream has not been imported yet", statusRef.Tag)
+					}
+					// status tags that are newer should be chosen over the spec tag
+					if generation < newest.Generation {
+						continue
```

Comment:
When the status ref's generation is larger than the generation we see in the spec tag, isn't that fine?  I don't understand what this block is protecting us from.  I guess I would expect something like "if `Items[0]` is too new (because it has too large a generation), and for some reason we are worried about that large generation (not clear to me why that would be), then walk `Items[1...]` to see if we can find an exact match, and use that".  Not the current "`continue`, I guess we want to silently ignore this `ref`".

Reply:
[This][1] is me stumbling around in the ImageStream data which lead to the bug.  I still don't understand this spec generation < newest status generation case, but it looks like there are an awful lot of those tags, including things like the CVO, that we need to get a functional release image:

```console
$ yaml2json <ocp.4.8.is.yaml | jq '.[].spec.tags[] | select(.name == "cluster-version-operator")'
{
  "annotations": null,
  "from": {
    "kind": "DockerImage",
    "name": "registry.ci.openshift.org/ocp/4.8@sha256:f1582d5e229f03ea2ee995e6813392695b85a7b7e6c539cacb49261aa9d9879c"
  },
  "generation": 2716,
  "importPolicy": {},
  "name": "cluster-version-operator",
  "referencePolicy": {
    "type": "Source"
  }
}
$ yaml2json <ocp.4.8.is.yaml | jq '.[].status.tags[] | select(.tag == "cluster-version-operator")'
{
  "items": [
    {
      "created": "2021-02-18T23:29:19Z",
      "dockerImageReference": "image-registry.openshift-image-registry.svc:5000/ocp/4.8@sha256:96124bee9c94ffcc36bd60348b2b6cadd235c501b91e972fbbda6e15160e916a",
      "generation": 3057,
      "image": "sha256:96124bee9c94ffcc36bd60348b2b6cadd235c501b91e972fbbda6e15160e916a"
    },
    {
      "created": "2021-02-16T23:23:12Z",
      "dockerImageReference": "image-registry.openshift-image-registry.svc:5000/ocp/4.8@sha256:ef8c687b2b64cf9f1c849ade80cddf014dd44be3111330ac23bfa9638860c5b3",
      "generation": 3057,
      "image": "sha256:ef8c687b2b64cf9f1c849ade80cddf014dd44be3111330ac23bfa9638860c5b3"
    },
    {
      "created": "2021-02-16T02:06:20Z",
      "dockerImageReference": "image-registry.openshift-image-registry.svc:5000/ocp/4.8@sha256:a7cac3991b20066042e8df874c0d6673d5ed82d8cf3d1a2d62f3f28de44c3aa3",
      "generation": 3057,
      "image": "sha256:a7cac3991b20066042e8df874c0d6673d5ed82d8cf3d1a2d62f3f28de44c3aa3"
    }
  ],
  "tag": "cluster-version-operator"
}
```

[1]: https://bugzilla.redhat.com/show_bug.cgi?id=1937020#c3

Reply:
Status tags are created by a) tagging in spec, b) pushes (which don't change spec).  At the time a status tag is created the current metadata.generation is taken.  Any time a spec tag is changed the tag gets the generation of metadata.generation.  If status tag generation > spec tag generation, someone pushed.  If status tag generation < spec tag generation, we're waiting to import that tag.  If spec tag generation == status tag generation, we've imported the spec tag.

Reply:
Everything besides item[0] is history.  So generation comparison is about waiting for import, picking the pushed tag, or taking the spec tag.

Reply:
So:

* `generation > newest.Generation`, we're waiting for an import, error out (as you are currently doing).
* `generation == newest.Generation`, we've imported a spec tag.  But we may also have had subsequent pushes, as long as nobody else has touched something to get `metadata.generation` bumped.  For example, see the `machine-os-content` entries in [my bug comment][1].  I expect we want to use the status tag, because it's at least as new as the spec tag.
 * `generation < newest.Generation`, someone has definitely pushed since the last time the spec tag was bumped.  Use the status tag.

Doesn't that mean we want to drop this `generation < newest.Generation` `continue` block this thread is attached to?

[1]: https://bugzilla.redhat.com/show_bug.cgi?id=1937020#c3

Reply:
Ah, I see the intention is to fall through to [the `status.tags` iterator][1] in the following block.  `continue` makes sense for that, but then I think it should be `<=` here to fall back to status for the `machine-os-content` pushed-since-the-last-spec-bump case.

[1]: https://github.com/openshift/oc/blob/be64dc73d95b407c2b0b36026fd0c38198b5be3e/pkg/cli/admin/release/new.go#L861-L866

Reply:
this can only be a digest, and so by design they would be equal.  if they’re equal, would prefer to emphasize we’re choosing the spec tag (by handling it here)

Reply:
also the handling of “not yet imported” is inside this loop, so <= would break that handling

Reply:
added a bit more comment